### PR TITLE
feat: Distillation Workflow for Whittle

### PR DESCRIPTION
**Pull request by HannanJaved**
### What does this implement/fix? Explain your changes.

- Added a distillation folder in whittle with a distill_student.py file - main distillation workflow
- Updated kd_loss in loss (CE loss was throwing some dimensionality errors)
- Added DistillArgs to args.py
- Added a test for the distillation loss to ensure the outputs are as expected

These files together should allow for a distillation workflow. (Issue #169)
Currently, the file initializes a random subnet of the teacher model as a student model. 
The code is adapted from pretrain_super_network.py

TODO:
Add a colab tutorial for distillation
Add a test for distillation training similar to test_pretrain_super_network

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the
terms of your choice.

### CODE REVIEW

File: whittle/distillation/distill.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+    teacher.eval()
+    
+    fabric.print(f"Teacher model loaded from {initial_checkpoint_dir} (not compiled)")
+    fabric.print(f"Teacher model has {compute_parameters(teacher):,} parameters")
+
+    # Get search space for creating different subnetworks
+    search_space = get_search_space(teacher_config)
+
+    sampler = RandomSampler(search_space, seed=seed)
+    random_config = sampler.sample()
```
Comment (gabikadlecova) [inline whittle/distillation/distill.py:258]: Can we also provide a path to a saved sub-network config? So that we can also distill into a previously chosen sub-network.

File: whittle/distillation/distill.py
Comment (aaronkl) [inline whittle/distillation/distill.py:1]: All the other workflows (e.g pretrain_super_network.py etc) are in the main folder, so I would suggest to move this one as well.

File: whittle/distillation/distill.py [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+    Arguments:
+        load_student_checkpoint: Whether to load the student model from a checkpoint.
+        out_dir: Directory in which to save checkpoints and logs. If running in a Lightning Studio Job, look for it in
+            /teamspace/jobs/<job-name>/share.
+        precision: The precision to use for finetuning. Determines a compatible precision setting by default.
+        initial_checkpoint_dir: Path to a checkpoint directory to initialize the teacher model from.
+        student_dir: Optional path to a directory to initialize the student model from.
+        data: Data-related arguments. If not provided, the default is ``litgpt.data.TinyStories``.
+        train: Training-related arguments. See ``litgpt.args.TrainArgs`` for details.
+        eval: Evaluation-related arguments. See ``litgpt.args.EvalArgs`` for details.
```
Comment (aaronkl) [inline whittle/distillation/distill.py:99]: Could you add distill here also to the docstring?

File: whittle/distillation/distill.py [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+    precision = precision or get_default_supported_precision(training=True)
+    num_devices = int(parse_devices(devices))
+    out_dir = init_out_dir(out_dir)
+    # in case the dataset requires the Tokenizer
+    tokenizer = Tokenizer(tokenizer_dir) if tokenizer_dir is not None else None
+
+    logger = choose_logger(
+        logger_name,
+        out_dir,
+        name=f"one-model-distill-experiments-{config.name}",
```
Comment (aaronkl) [inline whittle/distillation/distill.py:134]: could we rename this to for example "distill-{config.name}"

File: whittle/distillation/distill.py [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+        else:
+            student = GPT(teacher_config)
+
+    if load_student_checkpoint:
+        checkpoint = os.path.join(student_dir, "lit_model.pth")
+        ckpt = torch.load(checkpoint, map_location=fabric.device, weights_only=True)
+        student.load_state_dict(ckpt, strict=False)
+        fabric.print(f"Student model loaded from {student_dir} (not compiled)")
+    
+    if student_config is None:
```
Comment (aaronkl) [inline whittle/distillation/distill.py:254]: Instead of sampling a random sub-network, could we just have a flag (e.g `init_from_super_net`) that determines if we use a sub-network from the teacher network as student using the student config instead of a random config?

Comment (HannanJaved) [reply whittle/distillation/distill.py:254]: That is currently being checked by the student_dir argument. If no path is provided, a random subnet is sampled, otherwise a student config (and if available, checkpoint) is used.

File: whittle/distill.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+    )
+
+    if torch.cuda.is_available() and num_devices > 1:
+        check_nvlink_connectivity(fabric)
+
+    fabric.launch()
+
+    fabric.print(pprint.pformat(hparams))
+    if logger_name in ("tensorboard", "wandb"):
+        fabric.logger.log_hyperparams(hparams)
```
Comment (gabikadlecova) [inline whittle/distill.py:165]: We could also dump them into hyperparameters.yaml locally

File: whittle/distill.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        eval,
+        optimizer,
+        student_dir,
+    )
+
+def main(
+    fabric: L.Fabric,
+    devices: int,
+    seed: int = 42,
+    initial_checkpoint_dir: Path | None = None,
```
Comment (gabikadlecova) [inline whittle/distill.py:189]: I think None is not a valid default so we should have it as a positional args

File: whittle/distill.py [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+            "seed": seed,
+            "embed_dim": random_config["embed_dim"],
+            "mlp_ratio": random_config["mlp_ratio"],
+            "depth": random_config["depth"],
+            "parameter_count": param_count,
+            "reduction_ratio": param_count/compute_parameters(teacher)
+        }
+        student.config.n_embd = exp_config["embed_dim"]
+        student.config.intermediate_size = int(exp_config["mlp_ratio"] * exp_config["embed_dim"])
+        student.config.n_layer = exp_config["depth"]
```
Comment (gabikadlecova) [inline whittle/distill.py:306]: This looks a bit dangerous to overwrite it - I don't think we have it covered by tests. If you want the student to be a super-network, use `extract_sub_network`. Is there any other reason for setting it?

File: whittle/distill.py [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+        fabric.print(f"Student model set to a random subnetwork of the teacher model")
+        search_space = get_search_space(teacher_config)
+        sampler = RandomSampler(search_space, seed=seed)
+        random_config = sampler.sample()
+        subnetwork = {
+            "sub_network_n_embd": random_config["embed_dim"],
+            "sub_network_intermediate_size": int(random_config["mlp_ratio"] * random_config["embed_dim"]),
+            "sub_network_num_heads": teacher_config.n_head, # keep the same number of heads as teacher
+            "sub_network_n_layers": random_config["depth"]
+        }
```
Comment (gabikadlecova) [inline whittle/distill.py:262]: Actually I'd modify it or omit it from the script. Usually, you'd like to train a sub-network with a specific parameter count.

I propose either
a) sample a sub-network until it falls into a pre-defined parameter range (specify from - to in setup() args), or
b) move that logic into a separate file "get sub-network at param range" and remove this part

Comment (gabikadlecova) [reply whittle/distill.py:262]: @aaronkl  what do you think?

File: whittle/distill.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+    throughput = ThroughputMonitor(fabric, window_size=5)
+
+    with torch.device("meta"):
+        meta_model = GPT(student.config)
+        meta_model.set_sub_network(**{
+            "sub_network_n_embd": student.config.n_embd,
+            "sub_network_intermediate_size": student.config.intermediate_size,
+            "sub_network_num_heads": student.config.n_head,
+            "sub_network_n_layers": student.config.n_layer,
+        })
```
Comment (gabikadlecova) [inline whittle/distill.py:396]: If the student is a super-network, this is a bit superfluous (but probably not outright bad either).

**Review by gabikadlecova — CHANGES_REQUESTED** [review]
Overall, it looks good! I had some nitpicks for args, and one larger comment about student sub-network sampling

File: whittle/distillation/distill.py [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+        method='logits',
+        temperature=5,
+        alpha=0.7,
+    ),
+    eval: EvalArgs = EvalArgs(interval=1000, max_iters=100),
+    optimizer: str | dict = "AdamW",
+    devices: int | str = "auto",
+    num_nodes: int = 1,
+    tokenizer_dir: Path | None = None,
+    logger_name: Literal["wandb", "tensorboard", "csv"] = "wandb",
```
[the remark this answers is not in the document]
Comment (HannanJaved) [reply whittle/distillation/distill.py]: Yes, probably csv. Seen that as a default the most.

File: whittle/distill.py [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+from whittle.sampling.random_sampler import RandomSampler
+from whittle.pretrain_super_network import get_search_space
+from whittle.models.gpt.model import GPT
+from whittle.models.gpt.blocks import Block
+from whittle.args import DistillArgs
+from whittle.loss.kd_loss import DistillLoss
+from whittle.metrics.parameters import compute_parameters
+
+from jsonargparse import CLI
+from pathlib import Path
```
Comment (timurcarstensen) [inline whittle/distill.py:56]: Seems like the pre-commit hooks aren't catching the import sorting for some reason. Could you check whether this improves if you merge main and commit again?

**Review by gabikadlecova — APPROVED** [review]
I think it looks good now! I had only 2 small comments (out_dir name, docstring for default args). After making the tests pass it's ready to be merged

File: whittle/args.py [hunk 12]
```diff
@@ -18,6 +18,20 @@ class SearchArgs:
...
+
+    method: str = "logits"  # Only supports 'logits' for now
+    """Distillation method to use ('logits' or 'hidden_states')"""
+    temperature: float = 5
+    """Temperature for softening output probabilities. Higher values (>1) produce softer distributions,
+    emphasizing less confident predictions. Lower values (<1) make distributions sharper, focusing on
+    confident predictions."""
+    alpha: float = 0.5
+    """Weight balancing distillation loss vs cross-entropy loss. Values closer to 1 give more importance 
+    to matching teacher logits, while values closer to 0 prioritize true label prediction."""
```
[the remark this answers is not in the document]
Comment (HannanJaved) [reply whittle/args.py]: Changed; could you check if I understood it correctly :)

File: whittle/args.py [hunk 13]
```diff
@@ -18,6 +18,26 @@ class SearchArgs:
...
+        alpha: Weight balancing distillation loss vs cross-entropy loss. Values closer to 1 give more importance
+          to matching teacher logits, while values closer to 0 prioritize true label prediction.
+    """
+
+    method: str = "logits"
+    """Distillation method to use ('logits' or 'hidden_states')"""
+    temperature: float = 5
+    """Temperature for distillation"""
+    alpha: float = 0.5
+    """Weight for KL-Divergence (distillation) loss"""
```
Comment (timurcarstensen) [inline whittle/args.py:38]:
```suggestion
    method: str = "logits"
    temperature: float = 5
    alpha: float = 0.5
```

File: test/test_distill_loss.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    distill_weight = 0.5
+
+    loss_fn = DistillLoss(temperature, distill_weight)
+    loss = loss_fn(student_logits, labels, teacher_logits)
+
+    expected_loss = 0.6931
+    assert torch.isclose(loss, torch.tensor(expected_loss, dtype=loss.dtype), atol=1e-3)
+
+    loss.backward()
+    assert student_logits.grad is not None
```
Comment (timurcarstensen) [inline test/test_distill_loss.py:23]: no new line at end of file

File: whittle/args.py [hunk 15]
```diff
@@ -18,6 +18,23 @@ class SearchArgs:
...
+                confident predictions.
+        alpha: Weight balancing distillation loss vs cross-entropy loss. Values closer to 1 give more importance
+                to matching teacher logits, while values closer to 0 prioritize true label prediction.
+    """
+
+    method: str = "logits"
+    temperature: float = 5
+    alpha: float = 0.5
+
+
```
Comment (aaronkl) [inline whittle/args.py:37]:
```suggestion
@dataclass
```

File: whittle/distill.py [hunk 16]
```diff
@@ -0,0 +1,620 @@
...
+    logger_name: Literal["wandb", "tensorboard", "csv"] = "csv",
+    seed: int = 42,
+    min_ratio: float = 0.3,
+    max_ratio: float = 0.7,
+):
+    """Train a (random) subnet of the teacher model using knowledge distillation.
+
+    Arguments:
+        out_dir: Directory in which to save checkpoints and logs. If running in a Lightning Studio Job, look for it in
+            /teamspace/jobs/<job-name>/share.
```
Comment (timurcarstensen) [inline whittle/distill.py:89]: shall we remove the lightning studio reference here?

File: whittle/distill.py (hunk 11 above)
Comment (timurcarstensen) [reply whittle/distill.py]: Also: did you install the hooks? If you haven't, running `pre-commit install` and `pre-commit run --all-files` should fix the issues you're seeing in CI

File: whittle/loss/kd_loss.py [hunk 17]
```diff
@@ -51,16 +51,22 @@ def forward(self, outputs, labels, outputs_teacher) -> nn.Tensor:
...
+
         if outputs_teacher is not None and self.distillation_weight > 0:
             soft_target_loss = self.kldiv(
                 F.log_softmax(outputs / self.temperature, dim=1),
                 F.softmax(outputs_teacher / self.temperature, dim=1),
             ) * (self.temperature**2)
 
-        hard_target_loss = F.cross_entropy(outputs, labels, reduction="mean")
+        hard_target_loss = F.cross_entropy(
+            outputs.reshape(-1, outputs.size(-1)),
```
Comment (aaronkl) [inline whittle/loss/kd_loss.py:62]: Changing the shape here leads to a crash in the other unit test. Could we revert the change here and instead reshape the output in the distill.py

**Review by aaronkl — CHANGES_REQUESTED** [review]

File: whittle/distill.py (hunk 16 above)
Comment (HannanJaved) [reply whittle/distill.py:89]: removed it

**Review by aaronkl — APPROVED** [review]
